### PR TITLE
feat: Observed element hooks

### LINKED ISSUE #27961: [Docs]: add docs of usage for @fluenui/react-tabster
### Discussed in https://github.com/microsoft/fluentui/discussions/27883

<div type='discussions-op-text'>

<sup>Originally posted by **AkiraVoid** May 17, 2023</sup>

 the docs of @fluenui/react-tabster should be improved in case the users don't know how to use it.

### LINKED ISSUE #28030: [Feature]: focusAsync counterpart in @fluentui/react-components
### Library

React Components / v9 (@fluentui/react-components)

### Describe the feature that you would like added

Fluent 8 has [focusAsync utility](https://github.com/microsoft/fluentui/blob/e95ec0fee21e2ceb113d0678520d2bf6084b1495/packages/utilities/src/focus.ts#LL488C17-L488C27). However, there is no counterpart for it in Fluent 9. It would be great if we had similar utility. With tabster it's possible just by reusing [Observed API](https://tabster.io/docs/observed#requestfocus), with additional feature of being able to focus element which have not yet apperead in DOM. I was thinking about something like:

```ts
import { useTabsterAttributes } from "@fluentui/react-tabster";
import { createTabster, getObservedElement } from "tabster";
import type { Types } from "tabster";

export const useFocusElement = (
  name: string,
  timeout = 1000
): {
  attributes: Types.TabsterDOMAttribute;
  focus: () => Types.ObservedElementAsyncRequest<boolean>;
} => {
  const attributes = useTabsterAttributes({ observed: { names: [name] } });

  const tabsterCore = createTabster(window); // if this was part of @fluentui/react-tabster, we could call useTabter here instead

  const observedElement = getObservedElement(tabsterCore);
  return {
    attributes,
    focus: () => observedElement.requestFocus(name, timeout),
  };
};
```

What do you think?

### Have you discussed this feature with our team

_No response_

### Additional context

_No response_

### Validations

- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by ling1726**
Implements hooks that enable the use of `observedElement` API in react-tabster

Fixes #28030
Fixes #27961

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/0de1de9afea49edc454c6141050d5baaac609f18'>0de1de9afea49edc454c6141050d5baaac609f18</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/Avatar-mount_base.html">610</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/Avatar-mount.html">627</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/Button-mount_base.html">289</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/Button-mount.html">299</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/Field-mount_base.html">1056</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/Field-mount.html">1036</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/FluentProvider-mount_base.html">654</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/FluentProvider-mount.html">665</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/FluentProviderWithTheme-mount_base.html">81</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/FluentProviderWithTheme-mount.html">76</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">64</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">66</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">68</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">75</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/InfoButton-mount_base.html">9</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/InfoButton-mount.html">14</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/MakeStyles-mount_base.html">854</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/MakeStyles-mount.html">836</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/Persona-mount_base.html">1667</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/Persona-mount.html">1591</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/SpinButton-mount_base.html">1295</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28291/perf-test-react-components/SpinButton-mount.html">1305</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluentuiv9"/>

**Review by layershifter — CHANGES_REQUESTED** [review]
Please add usage documentation

**Review by bsunderhus — APPROVED** [review]

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useObservedElement/Default.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    }
+  };
+
+  return (
+    <div>
+      <Button onClick={onClick}>{mounted ? 'Reset' : 'Load and Focus'}</Button>
+      {mounted ? <Button {...attr}>LoadAndFocus</Button> : null}
+    </div>
+  );
+};
```
Comment (layershifter) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useObservedElement/Default.stories.tsx:32]: Made it a bit nicer:

```tsx
import * as React from 'react';
import {
  Button,
  useObservedElement,
  useFocusObserved,
  useId,
  makeStyles,
  shorthands,
  tokens,
  ToggleButton,
  Title3,
} from '@fluentui/react-components';

const useStyles = makeStyles({
  container: {
    display: 'flex',
    flexDirection: 'column',
    ...shorthands.gap('10px'),
  },
  card: {
    display: 'flex',
    flexDirection: 'column',
    ...shorthands.gap('10px'),

    backgroundColor: tokens.colorBrandBackground2,
    ...shorthands.borderRadius(tokens.borderRadiusMedium),
    ...shorthands.padding('10px'),
  },
});

export const Default = () => {
  const styles = useStyles();
  const observedName = useId('observed');

  const attributes = useObservedElement(observedName);
  const focus = useFocusObserved(observedName);

  const [mounted, setMounted] = React.useState(false);
  const [loading, setLoading] = React.useState(false);

  React.useEffect(() => {
    if (mounted) {
      focus();
    }
  }, [mounted, focus]);

  const onClick = () => {
    if (mounted) {
      setMounted(false);
      setLoading(false);
    } else {
      setLoading(true);
      setTimeout(() => {
        setMounted(true);
        setLoading(false);
      }, 1000);
    }
  };

  return (
    <div className={styles.container}>
      <div>
        <ToggleButton checked={mounted} disabledFocusable={loading} onClick={onClick}>
          {mounted ? 'Reset' : 'Load and Focus'}
        </ToggleButton>
      </div>

      {mounted ? (
        <div className={styles.card}>
          <Title3>Hello world!</Title3>
          <div>
            <Button {...attributes}>Focused on load</Button>
          </div>
        </div>
      ) : null}
    </div>
  );
};
```

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/FindPrevious.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      {...attributes}
+      className={styles.container}
+    >
+      <Button aria-label="Bold" icon={<TextBoldRegular />} />
+      <Button aria-label="Underline" icon={<TextUnderlineRegular />} />
+      <Button aria-label="Italic" icon={<TextItalicRegular />} />
+      <Button aria-label="Align Left" icon={<TextAlignLeftRegular />} />
+      <Button aria-label="Align Center" icon={<TextAlignCenterRegular />} />
+      <Button appearance="primary" aria-label="Align Right" icon={<TextAlignRightRegular />} />
+      <Button onClick={onClick}>Focus next</Button>
```
Comment (ling1726) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/FindPrevious.stories.tsx:47]:
```suggestion
      <Button onClick={onClick}>Focus previous</Button>
```

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useModalAttributes/Default.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  tokens,
+  Title2,
+} from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  dialog: {
+    position: 'fixed',
+    backgroundColor: tokens.colorNeutralBackground1,
+    ...shorthands.inset(0),
+    ...shorthands.padding(0),
```
Comment (layershifter) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useModalAttributes/Default.stories.tsx:17]:
```suggestion
```

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useModalAttributes/InertFocusTrap.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  tokens,
+  Title2,
+} from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  dialog: {
+    position: 'fixed',
+    backgroundColor: tokens.colorNeutralBackground1,
+    ...shorthands.inset(0),
+    ...shorthands.padding(0),
```
Comment (layershifter) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useModalAttributes/InertFocusTrap.stories.tsx:17]:
```suggestion
```

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useModalAttributes/Default.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
+import * as React from 'react';
```
Comment (layershifter) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useModalAttributes/Default.stories.tsx:1]: Small improvements:

```tsx
import * as React from 'react';
import {
  useModalAttributes,
  useFocusFinders,
  makeStyles,
  shorthands,
  Button,
  tokens,
  Title2,
} from '@fluentui/react-components';

const useStyles = makeStyles({
  dialog: {
    position: 'fixed',
    backgroundColor: tokens.colorNeutralBackground1,
    ...shorthands.inset(0),
    ...shorthands.padding('10px'),
    ...shorthands.margin('auto'),
    ...shorthands.borderRadius(tokens.borderRadiusMedium),
    ...shorthands.overflow('unset'),
    boxShadow: tokens.shadow16,
    width: '450px',
    height: '200px',
    display: 'flex',
    flexDirection: 'column',
  },
  title: {
    marginTop: 0,
    marginBottom: '10px',
  },
  paragraph: {
    marginTop: 0,
    marginBottom: '10px',
  },
  footer: {
    display: 'flex',
    marginTop: 'auto',
    justifyContent: 'end',
    ...shorthands.gap('5px'),
  },
});

export const Default = () => {
  const styles = useStyles();
  const [open, setOpen] = React.useState(false);

  const { triggerAttributes, modalAttributes } = useModalAttributes({ legacyTrapFocus: true, trapFocus: true });
  const { findFirstFocusable } = useFocusFinders();

  const triggerRef = React.useRef<HTMLButtonElement>(null);
  const dialogRef = React.useRef<HTMLDivElement>(null);

  const onClickTrigger = () => {
    setOpen(true);
  };

  const onClickClose = () => {
    setOpen(false);
    triggerRef.current?.focus();
  };

  const onDialogKeydown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    if (e.key === 'Escape') {
      setOpen(false);
      triggerRef.current?.focus();
    }
  };

  React.useEffect(() => {
    if (open && dialogRef.current) {
      findFirstFocusable(dialogRef.current)?.focus();
    }
  }, [open, findFirstFocusable]);

  return (
    <>
      <Button ref={triggerRef} {...triggerAttributes} onClick={onClickTrigger}>
        Open dialog
      </Button>
      {open && (
        /** FIXME ⚠️ This is a dialog for example purposes
                     use Dialog component in the application instead */
        <div
          onKeyDown={onDialogKeydown}
          ref={dialogRef}
          {...modalAttributes}
          aria-modal="true"
          role="dialog"
          className={styles.dialog}
          aria-label="Example dialog"
        >
          <Title2 as="h2" className={styles.title}>
            Example dialog
          </Title2>
          <p className={styles.paragraph}>Hello world!</p>
          <p className={styles.paragraph}>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et
            dolore magna aliqua.
          </p>
          <div className={styles.footer}>
            <Button appearance="primary">Action</Button>
            <Button onClick={onClickClose}>Close</Button>
          </div>
        </div>
      )}
    </>
  );
};

```

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/Default.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        <Button aria-label="Align Left" icon={<TextAlignLeftRegular />} />
+        <Button aria-label="Align Center" icon={<TextAlignCenterRegular />} />
+        <Button aria-label="Align Right" icon={<TextAlignRightRegular />} />
+        <Button aria-label="Copy" icon={<CopyRegular />} />
+        <Button aria-label="Cut" icon={<CutRegular />} />
+        <Button aria-label="Paste" icon={<ClipboardPasteRegular />} />
+      </div>
+    </>
+  );
+};
```
Comment (layershifter) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/Default.stories.tsx:66]:
```tsx
import * as React from 'react';
import {
  TextBoldRegular,
  TextUnderlineRegular,
  TextItalicRegular,
  TextAlignLeftRegular,
  TextAlignCenterRegular,
  TextAlignRightRegular,
  CopyRegular,
  ClipboardPasteRegular,
  CutRegular,
} from '@fluentui/react-icons';
import {
  Button,
  useArrowNavigationGroup,
  makeStyles,
  shorthands,
  useFocusFinders,
  Body1Stronger,
  tokens,
} from '@fluentui/react-components';

const useStyles = makeStyles({
  container: {
    display: 'flex',
    ...shorthands.gap('5px'),
    ...shorthands.padding('10px'),

    ...shorthands.border('2px', 'solid', tokens.colorBrandStroke1),
    borderBottomLeftRadius: tokens.borderRadiusMedium,
    borderBottomRightRadius: tokens.borderRadiusMedium,
  },
  note: {
    display: 'block',
    backgroundColor: tokens.colorBrandStroke1,
    color: tokens.colorNeutralForegroundOnBrand,

    ...shorthands.padding('4px'),
    borderTopLeftRadius: tokens.borderRadiusMedium,
    borderTopRightRadius: tokens.borderRadiusMedium,
  },
});

export const Default = () => {
  const styles = useStyles();
  const { findAllFocusable } = useFocusFinders();
  const [count, setCount] = React.useState(0);
  const toolbarRef = React.useRef<HTMLDivElement>(null);
  const attributes = useArrowNavigationGroup({ axis: 'horizontal' });

  React.useEffect(() => {
    if (toolbarRef.current) {
      setCount(findAllFocusable(toolbarRef.current).length);
    }
  }, [findAllFocusable]);

  return (
    <>
      <div className={styles.note}>
        <Body1Stronger>{count} focusable elements below</Body1Stronger>
      </div>
      <div
        ref={toolbarRef}
        aria-label="Editor toolbar example"
        role="toolbar"
        {...attributes}
        className={styles.container}
      >
        <Button aria-label="Bold" icon={<TextBoldRegular />} />
        <Button aria-label="Underline" icon={<TextUnderlineRegular />} />
        <Button aria-label="Italic" icon={<TextItalicRegular />} />
        <Button aria-label="Align Left" icon={<TextAlignLeftRegular />} />
        <Button aria-label="Align Center" icon={<TextAlignCenterRegular />} />
        <Button aria-label="Align Right" icon={<TextAlignRightRegular />} />
        <Button aria-label="Copy" disabled icon={<CopyRegular />} />
        <Button aria-label="Cut" disabled icon={<CutRegular />} />
        <Button aria-label="Paste" disabled icon={<ClipboardPasteRegular />} />
      </div>
    </>
  );
};
```

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/FindFirst.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      </div>
+      <div
+        ref={toolbarRef}
+        aria-label="Editor toolbar example"
+        role="toolbar"
+        {...attributes}
+        className={styles.container}
+      >
+        <Button appearance="primary" aria-label="Bold" icon={<TextBoldRegular />} />
+        <Button aria-label="Underline" icon={<TextUnderlineRegular />} />
```
Comment (layershifter) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/FindFirst.stories.tsx:48]:
```suggestion
        <Button disabled aria-label="Bold" icon={<TextBoldRegular />} />
        <Button appearance="primary" aria-label="Underline" icon={<TextUnderlineRegular />} />
```

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/FindLast.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  ClipboardPasteRegular,
+  CutRegular,
+} from '@fluentui/react-icons';
+import { Button, useArrowNavigationGroup, makeStyles, shorthands, useFocusFinders } from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  container: {
+    marginTop: '5px',
+    display: 'flex',
+    ...shorthands.gap('5px'),
```
Comment (layershifter) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/FindLast.stories.tsx:19]:
```suggestion
    ...shorthands.gap('5px'),
    ...shorthands.padding('10px'),

    ...shorthands.border('2px', 'solid', tokens.colorBrandStroke1),
    ...shorthands.borderRadius(tokens.borderRadiusMedium),
```

File: packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/FindLast.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      >
+        <Button aria-label="Bold" icon={<TextBoldRegular />} />
+        <Button aria-label="Underline" icon={<TextUnderlineRegular />} />
+        <Button aria-label="Italic" icon={<TextItalicRegular />} />
+        <Button aria-label="Align Left" icon={<TextAlignLeftRegular />} />
+        <Button aria-label="Align Center" icon={<TextAlignCenterRegular />} />
+        <Button aria-label="Align Right" icon={<TextAlignRightRegular />} />
+        <Button aria-label="Copy" icon={<CopyRegular />} />
+        <Button aria-label="Cut" icon={<CutRegular />} />
+        <Button appearance="primary" aria-label="Paste" icon={<ClipboardPasteRegular />} />
```
Comment (layershifter) [inline packages/react-components/react-components/stories/Concepts/FocusManagement/useFocusFinders/FindLast.stories.tsx:55]:
```suggestion
        <Button appearance="primary" aria-label="Cut" icon={<CutRegular />} />
        <Button disabled aria-label="Paste" icon={<ClipboardPasteRegular />} />
```

**Review by layershifter — APPROVED** [review]

**Pull request merged**